pkg/tekton/metapipeline/metaclient/metaclient.go

```diff
@@ -0,0 +1,324 @@
+	f := o.GetFactory()
+
+	tektonClient, _, err := f.CreateTektonClient()
+	if err != nil {
+		return nil, nil, nil, "", errors.Wrap(err, "unable to create Tekton client")
+	}
+
+	jxClient, _, err := f.CreateJXClient()
+	if err != nil {
+		return nil, nil, nil, "", errors.Wrap(err, "unable to create JX client")
+	}
+
+	kubeClient, ns, err := f.CreateKubeClient()
+	if err != nil {
+		return nil, nil, nil, "", errors.Wrap(err, "unable to create Kube client")
+	}
+	ns, _, err = kube.GetDevNamespace(kubeClient, ns)
+	if err != nil {
+		return nil, nil, nil, "", errors.Wrap(err, "unable to find the dev namespace")
+	}
+	return tektonClient, jxClient, kubeClient, ns, nil
+}
+
+// determineBranchIdentifier finds a identifier for the branch which is build by the specified pull ref.
+// The name is either an actual git branch name (eg master) or a synthetic names  like PR-<number> for single pull requests
+// or 'batch' for Prow batch builds.
+// The method makes the decision purely based on the Prow pull ref. At this stage we don't have the full ProwJov spec
+// available.
+func (o *MetaClient) determineBranchIdentifier(pullRef prow.PullRefs) string {
+	prCount := len(pullRef.ToMerge)
+	var branch string
+	switch prCount {
+	case 0:
+		{
+			// no pull requests to merge, taking base branch name as identifier
+			branch = pullRef.BaseBranch
+		}
+	case 1:
+		{
+			// single pull request, create synthetic PR identifier
+			for k := range pullRef.ToMerge {
+				branch = fmt.Sprintf("PR-%s", k)
+				break
+			}
+		}
+	default:
+		{
+			branch = "batch"
+		}
+	}
+	log.Logger().Debugf("branch identifier for pull ref '%s' : '%s'", pullRef.String(), branch)
+	return branch
+}
+
+func (o *MetaClient) determinePipelineKind(pullRef prow.PullRefs) string {
+	var kind string
+
+	prCount := len(pullRef.ToMerge)
+	if prCount > 0 {
+		kind = jenkinsfile.PipelineKindPullRequest
+	} else {
+		kind = jenkinsfile.PipelineKindRelease
+	}
+	log.Logger().Debugf("pipeline kind for pull ref '%s' : '%s'", pullRef.String(), kind)
+	return kind
+}
+
+func (o *MetaClient) gitCloneVersionStream(jxClient jxclient.Interface, ns string) (string, error) {
+	dir, err := ioutil.TempDir("", "jx-version-repo-")
+	if err != nil {
+		return dir, err
+	}
+
+	devEnv, err := kube.GetDevEnvironment(jxClient, ns)
+	if err != nil {
+		return dir, err
+	}
+	teamSettings := devEnv.Spec.TeamSettings
+	url := teamSettings.VersionStreamURL
+	ref := teamSettings.VersionStreamRef
+	if url == "" {
+		url = config.DefaultVersionsURL
+	}
+	if ref == "" {
+		ref = config.DefaultVersionsRef
+	}
+	log.Logger().WithField("url", url).WithField("ref", ref).Infof("shallow cloning version stream")
+	err = o.Git().ShallowClone(dir, url, ref, "")
+	return dir, err
+}
+
+// Git returns the git client
+func (o *MetaClient) Git() gits.Gitter {
+	if o.git == nil {
+		o.git = gits.NewGitCLI()
+	}
+	return o.git
+}
+
+// SetGit sets the git client
+func (o *MetaClient) SetGit(git gits.Gitter) {
```

Comment:
ditto

Reply:
I left this one public so we can create test cases with fake gitters in other packages

Reply:
lemme make that one private until we need it